### DISCUSSION TITLE
Un-using

### DISCUSSION THREAD

**Discussion by wrexbe** [dotnet/csharplang#7303] [Language Ideas]
So my project currently has System.Text.Json, and Newtonsoft which is coming indirectly from a Nuget I am using. The IDE is being annoying, by suggesting Newtonsoft all the time. This is probably more of an IDE thing, but in the project, or in C# I want the ability to globally say not to use anything from a namespace, the opposite of a using. This would improve "Add missing references", and also add a guard against someone using a similarly named type by accident.

**Comment by 333fred** [comment]
@wrexbe you might be interested in the [Banned API Analyzers](https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/BannedApiAnalyzers.Help.md). It won't filter things out of completion, but it will give you an error if you use any of the types you specify.

**Reply by wrexbe** [reply]
Maybe a way to have the IDE detect when that analyzer is being used, and to suppress suggestions. Or maybe a way for analyzers to suppress/modify suggestions themselves.

**Comment by BrysAdamed** [comment]
I was recently been thinking about something like this. It would be nice if we could "ban" or "exclude" some namespaces, similarly to the global usings.
It could work like this:

```c#
exclude using System.Windows.Input; // ban whole namespace
exclude using System.Windows.Input.ICommand; // ban single class
```

Or in csproj :

```xml
<ItemGroup>
    <Using Exclude="System.Windows.Input"/>
</ItemGroup>
```

In my case i am writing apps with CQRS and we have the `ICommand` interface, and the IDE cannot automatically provide me with a correct using statement, because there also is a `System.Windows.Input.ICommand`.

**Comment by jnm2** [comment]
If your compilation doesn't need to directly reference any of the types from Newtonsoft.Json, you can add this to your .csproj to remove Newtonsoft.Json.dll as a compiler reference, while keeping Newtonsoft.Json.dll in your build output since the library you're using needs it:

```xml
  <Target Name="RemoveCompileTimeReferences" AfterTargets="ResolvePackageAssets" BeforeTargets="ResolveLockFileReferences">
    <ItemGroup>
      <ResolvedCompileFileDefinitions Remove="@(ResolvedCompileFileDefinitions)" Condition="'%(NuGetPackageId)' == 'Newtonsoft.Json'" />
    </ItemGroup>
  </Target>
  ```
